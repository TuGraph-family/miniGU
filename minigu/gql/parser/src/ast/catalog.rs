//! AST definitions for *catalog-modifying statements*.

use super::{
    CallProcedureStatement, CatalogObjectRef, GraphElementType, GraphExpr, GraphTypeRef, SchemaPath,
};
use crate::macros::base;
use crate::span::{OptSpanned, Spanned, VecSpanned};

pub type LinearCatalogModifyingStatement = VecSpanned<CatalogModifyingStatement>;

#[apply(base)]
pub enum CatalogModifyingStatement {
    Call(CallProcedureStatement),
    CreateSchema(CreateSchemaStatement),
    DropSchema(DropSchemaStatement),
    CreateGraph(CreateGraphStatement),
    DropGraph(DropGraphStatement),
    CreateGraphType(CreateGraphTypeStatement),
    DropGraphType(DropGraphTypeStatement),
}

#[apply(base)]
pub struct CreateSchemaStatement {
    pub path: Spanned<SchemaPath>,
    pub if_not_exists: bool,
}

#[apply(base)]
pub struct DropSchemaStatement {
    pub path: Spanned<SchemaPath>,
    pub if_exists: bool,
}

#[apply(base)]
pub struct CreateGraphStatement {
    pub path: Spanned<CatalogObjectRef>,
    pub kind: Spanned<CreateGraphOrGraphTypeStatementKind>,
    pub graph_type: Spanned<OfGraphType>,
    pub source: OptSpanned<GraphExpr>,
}

#[apply(base)]
pub enum CreateGraphOrGraphTypeStatementKind {
    Create,
    CreateIfNotExists,
    CreateOrReplace,
}

#[apply(base)]
pub enum OfGraphType {
    Like(Spanned<GraphExpr>),
    Ref(Spanned<GraphTypeRef>),
    Nested(VecSpanned<GraphElementType>),
    Any,
}

#[apply(base)]
pub struct DropGraphStatement {
    pub path: Spanned<CatalogObjectRef>,
    pub if_exists: bool,
}

#[apply(base)]
pub struct DropGraphTypeStatement {
    pub path: Spanned<CatalogObjectRef>,
    pub if_exists: bool,
}

#[apply(base)]
pub struct CreateGraphTypeStatement {
    pub path: Spanned<CatalogObjectRef>,
    pub kind: Spanned<CreateGraphOrGraphTypeStatementKind>,
    pub source: Spanned<GraphTypeSource>,
}

#[apply(base)]
pub enum GraphTypeSource {
    Copy(Spanned<GraphTypeRef>),
    Like(Spanned<GraphExpr>),
    Nested(VecSpanned<GraphElementType>),
}
