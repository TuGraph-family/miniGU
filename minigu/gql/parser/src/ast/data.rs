//! AST definitions for *data-modifying statements*.

use crate::macros::base;

#[apply(base)]
pub struct LinearDataModifyingStatement {}
