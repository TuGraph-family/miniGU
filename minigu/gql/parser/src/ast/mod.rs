pub mod catalog;
pub mod common;
pub mod data;
pub mod lexical;
pub mod object_expr;
pub mod object_ref;
pub mod predicate;
pub mod procedure_call;
pub mod procedure_spec;
pub mod program;
pub mod query;
pub mod session;
pub mod transaction;
pub mod type_element;
pub mod value_expr;
pub mod variable;

pub use catalog::*;
pub use common::*;
pub use data::*;
pub use lexical::*;
pub use object_expr::*;
pub use object_ref::*;
pub use predicate::*;
pub use procedure_call::*;
pub use procedure_spec::*;
pub use program::*;
pub use query::*;
pub use session::*;
pub use transaction::*;
pub use type_element::*;
pub use value_expr::*;
pub use variable::*;
