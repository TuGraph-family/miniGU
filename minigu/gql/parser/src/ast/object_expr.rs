//! AST definitions for *object expressions*.

use super::{Expr, GraphRef, Ident};
use crate::macros::base;
use crate::span::Spanned;

#[apply(base)]
pub enum GraphExpr {
    Name(Ident),
    Ref(GraphRef),
    Object(ObjectExpr),
    Current,
}

#[apply(base)]
pub enum ObjectExpr {
    Variable(Spanned<Expr>),
    Expr(Spanned<Expr>),
}
