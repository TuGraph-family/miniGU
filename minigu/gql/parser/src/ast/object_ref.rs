//! AST definitions for *object references*.

use super::Ident;
use crate::macros::base;
use crate::span::{OptSpanned, VecSpanned};

#[apply(base)]
pub enum PredefinedSchemaRef {
    Home,
    Current,
}

#[apply(base)]
pub enum SchemaRef {
    Absolute(SchemaPath),
    Relative(SchemaPath),
    Predefined(PredefinedSchemaRef),
    Parameter(Ident),
}

pub type SchemaPath = VecSpanned<SchemaPathSegment>;

#[apply(base)]
pub enum SchemaPathSegment {
    /// A named object (schema or directory), e.g., `a`.
    Name(Ident),
    /// Parent directory, e.g., `..`.
    Parent,
}

#[apply(base)]
pub enum GraphRef {
    Name(Ident),
    Parameter(Ident),
    Ref(CatalogObjectRef),
    Home,
}

#[apply(base)]
pub enum ProcedureRef {
    Ref(CatalogObjectRef),
    Parameter(Ident),
}

#[apply(base)]
pub enum GraphTypeRef {
    Ref(CatalogObjectRef),
    Parameter(Ident),
}

#[apply(base)]
pub enum BindingTableRef {
    Name(Ident),
    Ref(CatalogObjectRef),
    Parameter(Ident),
}

#[apply(base)]
pub struct CatalogObjectRef {
    pub schema: OptSpanned<SchemaRef>,
    pub objects: VecSpanned<Ident>,
}
