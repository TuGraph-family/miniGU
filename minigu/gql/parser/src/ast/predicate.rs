//! AST definitions for *predicates*.

use super::{Expr, Ident, LabelExpr, MatchStatement, Procedure, ValueType};
use crate::macros::base;
use crate::span::{BoxSpanned, Spanned, VecSpanned};

#[apply(base)]
pub enum Predicate {
    Exists(Exists),
    Null {
        expr: BoxSpanned<Expr>,
        not: bool,
    },
    Typed {
        expr: BoxSpanned<Expr>,
        value_type: BoxSpanned<ValueType>,
        not: bool,
    },
    Directed {
        edge: Spanned<Ident>,
        not: bool,
    },
    Labeled {
        element: Spanned<Ident>,
        label: Spanned<LabelExpr>,
    },
    SrcOf {
        node: Spanned<Ident>,
        edge: Spanned<Ident>,
        not: bool,
    },
    DstOf {
        node: Spanned<Ident>,
        edge: Spanned<Ident>,
        not: bool,
    },
    AllDifferent(VecSpanned<Ident>),
    Same(VecSpanned<Ident>),
    PropertyExists {
        element: Spanned<Ident>,
        property: Spanned<Ident>,
    },
}

#[apply(base)]
pub enum Exists {
    Pattern,
    Match(VecSpanned<MatchStatement>),
    Nested(BoxSpanned<Procedure>),
}
