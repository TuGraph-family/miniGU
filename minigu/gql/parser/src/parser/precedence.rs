use core::marker::PhantomData;

use winnow::Parser;
use winnow::combinator::opt;
use winnow::error::ParserError;
use winnow::stream::{Stream, StreamIsPartial};

/// Specifies the associativity of an (infix) operator.
#[derive(Debug, Clone, Copy)]
pub(super) enum Assoc {
    Left,
    Right,
    Neither,
}

pub(super) type Precedence = i64;

/// An operator precedence parser based on the [Pratt algorithm](https://en.wikipedia.org/wiki/Operator-precedence_parser).
/// The implementation is adapted from https://github.com/winnow-rs/winnow/pull/614.
#[allow(clippy::too_many_arguments)]
pub(super) fn precedence<
    I,
    E,
    Operand,
    PrefixOp,
    SuffixOp,
    InfixOp,
    ParseOperand,
    ParsePrefixOp,
    ParseSuffixOp,
    ParseInfixOp,
    FoldPrefix,
    FoldSuffix,
    FoldInfix,
>(
    init: Precedence,
    parse_operand: ParseOperand,
    parse_prefix: ParsePrefixOp,
    parse_suffix: ParseSuffixOp,
    parse_infix: ParseInfixOp,
    fold_prefix: FoldPrefix,
    fold_suffix: FoldSuffix,
    fold_infix: FoldInfix,
) -> impl Parser<I, Operand, E>
where
    I: Stream + StreamIsPartial,
    ParseOperand: Parser<I, Operand, E>,
    ParsePrefixOp: Parser<I, (Precedence, PrefixOp), E>,
    ParseSuffixOp: Parser<I, (Precedence, SuffixOp), E>,
    ParseInfixOp: Parser<I, (Assoc, Precedence, InfixOp), E>,
    FoldPrefix: FnMut(PrefixOp, Operand) -> Result<Operand, E>,
    FoldSuffix: FnMut(Operand, SuffixOp) -> Result<Operand, E>,
    FoldInfix: FnMut(Operand, InfixOp, Operand) -> Result<Operand, E>,
    E: ParserError<I>,
{
    let mut config = PrattConfig {
        parse_operand,
        parse_prefix,
        parse_suffix,
        parse_infix,
        fold_prefix,
        fold_suffix,
        fold_infix,
        _marker: PhantomData,
    };
    move |i: &mut _| config.precedence_impl(i, init)
}

struct PrattConfig<
    I,
    E,
    Operand,
    PrefixOp,
    SuffixOp,
    InfixOp,
    ParseOperand,
    ParsePrefixOp,
    ParseSuffixOp,
    ParseInfixOp,
    FoldPrefix,
    FoldSuffix,
    FoldInfix,
> {
    parse_operand: ParseOperand,
    parse_prefix: ParsePrefixOp,
    parse_suffix: ParseSuffixOp,
    parse_infix: ParseInfixOp,
    fold_prefix: FoldPrefix,
    fold_suffix: FoldSuffix,
    fold_infix: FoldInfix,
    _marker: PhantomData<(I, E, Operand, PrefixOp, SuffixOp, InfixOp)>,
}

impl<
    I,
    E,
    Operand,
    PrefixOp,
    SuffixOp,
    InfixOp,
    ParseOperand,
    ParsePrefixOp,
    ParseSuffixOp,
    ParseInfixOp,
    FoldPrefix,
    FoldSuffix,
    FoldInfix,
>
    PrattConfig<
        I,
        E,
        Operand,
        PrefixOp,
        SuffixOp,
        InfixOp,
        ParseOperand,
        ParsePrefixOp,
        ParseSuffixOp,
        ParseInfixOp,
        FoldPrefix,
        FoldSuffix,
        FoldInfix,
    >
where
    I: Stream + StreamIsPartial,
    ParseOperand: Parser<I, Operand, E>,
    ParsePrefixOp: Parser<I, (Precedence, PrefixOp), E>,
    ParseSuffixOp: Parser<I, (Precedence, SuffixOp), E>,
    ParseInfixOp: Parser<I, (Assoc, Precedence, InfixOp), E>,
    FoldPrefix: FnMut(PrefixOp, Operand) -> Result<Operand, E>,
    FoldSuffix: FnMut(Operand, SuffixOp) -> Result<Operand, E>,
    FoldInfix: FnMut(Operand, InfixOp, Operand) -> Result<Operand, E>,
    E: ParserError<I>,
{
    fn precedence_impl(&mut self, i: &mut I, min_power: Precedence) -> Result<Operand, E> {
        let operand = opt(self.parse_operand.by_ref()).parse_next(i)?;
        let mut operand = if let Some(operand) = operand {
            operand
        } else {
            let len = i.eof_offset();
            let (power, prefix_op) = self.parse_prefix.parse_next(i)?;
            if i.eof_offset() == len {
                return Err(E::assert(i, "`prefix` parsers must always consume"));
            }
            let operand = self.precedence_impl(i, power)?;
            (self.fold_prefix)(prefix_op, operand)?
        };

        let mut prev_op_is_neither = None;
        while i.eof_offset() > 0 {
            let start = i.checkpoint();
            if let Some((power, suffix_op)) = opt(self.parse_suffix.by_ref()).parse_next(i)? {
                if power < min_power {
                    i.reset(&start);
                    break;
                }
                operand = (self.fold_suffix)(operand, suffix_op)?;
                continue;
            }

            let start = i.checkpoint();
            let parse_result = opt(self.parse_infix.by_ref()).parse_next(i)?;
            if let Some((assoc, power, infix_op)) = parse_result {
                let mut is_neither = None;
                let (lpower, rpower) = match assoc {
                    Assoc::Left => (power, power + 1),
                    Assoc::Right => (power, power - 1),
                    Assoc::Neither => {
                        is_neither = Some(power);
                        (power, power + 1)
                    }
                };
                if lpower < min_power || prev_op_is_neither.is_some_and(|p| lpower == p) {
                    i.reset(&start);
                    break;
                }
                prev_op_is_neither = is_neither;
                let rhs = self.precedence_impl(i, rpower)?;
                operand = (self.fold_infix)(operand, infix_op, rhs)?;
                continue;
            }
            break;
        }
        Ok(operand)
    }
}

#[cfg(test)]
mod tests {
    use core::str::FromStr;

    use winnow::ModalResult;
    use winnow::ascii::{digit1, multispace0};
    use winnow::combinator::{delimited, dispatch, empty, fail, peek};
    use winnow::token::any;

    use super::*;

    fn operand(i: &mut &str) -> ModalResult<i64> {
        digit1.try_map(i64::from_str).parse_next(i)
    }

    fn parenthesized(i: &mut &str) -> ModalResult<i64> {
        delimited('(', expr, ')').parse_next(i)
    }

    fn expr(i: &mut &str) -> ModalResult<i64> {
        precedence(
            0,
            delimited(
                multispace0,
                dispatch! {peek(any);
                    '(' => parenthesized,
                    '0'..='9' => operand,
                    _ => fail
                },
                multispace0,
            ),
            delimited(
                multispace0,
                dispatch! {any;
                    '+' => empty.value((2, '+')),
                    '-' => empty.value((2, '-')),
                    _ => fail
                },
                multispace0,
            ),
            delimited(
                multispace0,
                dispatch! {any;
                    '!' => empty.value((3, '!')),
                    _ => fail
                },
                multispace0,
            ),
            delimited(
                multispace0,
                dispatch! {any;
                    '+' => empty.value((Assoc::Left, 0, '+')),
                    '-' => empty.value((Assoc::Left, 0, '-')),
                    '*' => empty.value((Assoc::Left, 1, '*')),
                    '/' => empty.value((Assoc::Left, 1, '/')),
                    _ => fail
                },
                multispace0,
            ),
            |op, a| match op {
                '+' => Ok(a),
                '-' => Ok(-a),
                _ => unreachable!(),
            },
            |a, op| match op {
                '!' => Ok((1..=a).product()),
                _ => unreachable!(),
            },
            |a, op, b| match op {
                '+' => Ok(a + b),
                '-' => Ok(a - b),
                '*' => Ok(a * b),
                '/' => Ok(a / b),
                _ => unreachable!(),
            },
        )
        .parse_next(i)
    }

    #[test]
    fn test_parse_operand() {
        assert_eq!(operand.parse("123").unwrap(), 123);
    }

    #[test]
    fn test_simple_expr() {
        assert_eq!(expr.parse("1 +  3* 4 /2 ").unwrap(), 7);
    }

    #[test]
    fn test_complex_expr() {
        assert_eq!(expr.parse("3- (+ 1 * - (3 +4) + 10)!").unwrap(), -3);
    }
}
