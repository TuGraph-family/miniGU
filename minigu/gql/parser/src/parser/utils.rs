use winnow::Parser;
use winnow::stream::{Location, Stream};

use crate::span::Spanned;

/// A helper macro for executing a parser on an input string. **This should only be used in unit
/// tests!**
#[cfg(all(test, feature = "serde"))]
macro_rules! parse {
    ($parser:expr, $input:literal) => {{
        let input = $input;
        $crate::parser::token::tokenize(input)
            .ok()
            .and_then(|tokens| {
                let stream = crate::parser::token::build_token_stream(&tokens, Default::default());
                $parser.parse(stream).ok()
            })
    }};
}

#[cfg(all(test, feature = "serde"))]
pub(super) use parse;

/// A helper trait for wrapping the output in [`Spanned`].
pub(super) trait ToSpanned<I, O, E>: Parser<I, O, E> {
    /// Wraps the output in [`Spanned`].
    #[inline(always)]
    fn spanned(self) -> impl Parser<I, Spanned<O>, E>
    where
        Self: Sized,
        I: Stream + Location,
    {
        self.with_span().map(|(inner, span)| Spanned(inner, span))
    }
}

/// A helper trait for operating on the value wrapped in [`Spanned`].
pub(super) trait SpannedParserExt<I, O, E>: Parser<I, Spanned<O>, E> {
    /// Unwraps the output from [`Spanned`].
    #[inline(always)]
    fn unspanned(self) -> impl Parser<I, O, E>
    where
        Self: Sized,
    {
        self.map(|spanned| spanned.0)
    }

    /// Maps the inner value of the output while keeping the span unchanged.
    #[inline(always)]
    fn map_inner<F, O2>(self, mut f: F) -> impl Parser<I, Spanned<O2>, E>
    where
        Self: Sized,
        F: FnMut(O) -> O2,
    {
        self.map(move |Spanned(inner, span)| Spanned(f(inner), span))
    }

    /// Updates the span of the output while keeping the inner value unchanged.
    #[inline(always)]
    fn update_span(self) -> impl Parser<I, Spanned<O>, E>
    where
        Self: Sized,
        I: Stream + Location,
    {
        self.with_span().map(|(mut output, new_span)| {
            output.1 = new_span;
            output
        })
    }
}

impl<I, O, E, P> ToSpanned<I, O, E> for P where P: Parser<I, O, E> {}

impl<I, O, E, P> SpannedParserExt<I, O, E> for P where P: Parser<I, Spanned<O>, E> {}

/// A helper macro for defining parser aliases.
macro_rules! def_parser_alias {
    ($name:ident, $parser:ident, $output:ty) => {
        #[inline(always)]
        pub fn $name(
            input: &mut $crate::parser::token::TokenStream,
        ) -> winnow::ModalResult<$output> {
            $parser(input)
        }
    };
}

pub(super) use def_parser_alias;
