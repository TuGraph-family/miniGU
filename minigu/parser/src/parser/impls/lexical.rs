use winnow::combinator::{cut_err, dispatch, empty, fail, peek};
use winnow::token::one_of;
use winnow::{ModalResult, Parser};

use crate::ast::{
    BooleanLiteral, Ident, Literal, StringLiteral, StringLiteralKind, UnsignedInteger,
    UnsignedIntegerKind, UnsignedNumericLiteral,
};
use crate::lexer::TokenKind;
use crate::parser::token::{Token, TokenStream, any};
use crate::parser::utils::{SpannedParserExt, ToSpanned, def_parser_alias};
use crate::span::Spanned;

pub fn regular_identifier(input: &mut TokenStream) -> ModalResult<Spanned<Ident>> {
    winnow::token::any
        .verify_map(|token: &Token| match &token.kind {
            &TokenKind::RegularIdentifier(name) => Some(name.into()),
            kind if kind.is_non_reserved_word() => Some(token.slice.into()),
            _ => None,
        })
        .spanned()
        .parse_next(input)
}

pub fn delimited_identifier(input: &mut TokenStream) -> ModalResult<Spanned<Ident>> {
    dispatch! {any;
        TokenKind::AccentQuoted(quoted) | TokenKind::DoubleQuoted(quoted) => {
            cut_err(empty.verify_map(|_| quoted.unescape()))
        },
        _ => fail,
    }
    .spanned()
    .parse_next(input)
}

pub fn identifier(input: &mut TokenStream) -> ModalResult<Spanned<Ident>> {
    dispatch! {peek(any);
        TokenKind::RegularIdentifier(_) => regular_identifier,
        kind if kind.is_non_reserved_word() => regular_identifier,
        TokenKind::AccentQuoted(_) | TokenKind::DoubleQuoted(_) => delimited_identifier,
        _ => fail
    }
    .parse_next(input)
}

def_parser_alias!(object_name, identifier, Spanned<Ident>);
def_parser_alias!(
    object_name_or_binding_variable,
    regular_identifier,
    Spanned<Ident>
);
def_parser_alias!(directory_name, identifier, Spanned<Ident>);
def_parser_alias!(schema_name, identifier, Spanned<Ident>);
def_parser_alias!(graph_name, identifier, Spanned<Ident>);
def_parser_alias!(delimited_graph_name, delimited_identifier, Spanned<Ident>);
def_parser_alias!(graph_type_name, identifier, Spanned<Ident>);
def_parser_alias!(node_type_name, identifier, Spanned<Ident>);
def_parser_alias!(edge_type_name, identifier, Spanned<Ident>);
def_parser_alias!(binding_table_name, identifier, Spanned<Ident>);
def_parser_alias!(
    delimited_binding_table_name,
    delimited_identifier,
    Spanned<Ident>
);
def_parser_alias!(procedure_name, identifier, Spanned<Ident>);
def_parser_alias!(label_name, identifier, Spanned<Ident>);
def_parser_alias!(property_name, identifier, Spanned<Ident>);
def_parser_alias!(field_name, identifier, Spanned<Ident>);
def_parser_alias!(element_variable, binding_variable, Spanned<Ident>);
def_parser_alias!(path_variable, binding_variable, Spanned<Ident>);
def_parser_alias!(subpath_variable, regular_identifier, Spanned<Ident>);
def_parser_alias!(binding_variable, regular_identifier, Spanned<Ident>);

pub fn unsigned_literal(input: &mut TokenStream) -> ModalResult<Spanned<Literal>> {
    dispatch! {peek(any);
        TokenKind::True | TokenKind::False | TokenKind::Unknown => {
            boolean_literal.map_inner(Literal::Boolean)
        },
        TokenKind::Null => TokenKind::Null.value(Literal::Null).spanned(),
        TokenKind::DoubleQuoted(_) | TokenKind::SingleQuoted(_) => {
            character_string_literal.map_inner(Literal::String)
        },
        kind if kind.is_prefix_of_numeric_literal() => {
            unsigned_numeric_literal.map_inner(Literal::Numeric)
        },
        _ => fail,
    }
    .parse_next(input)
}

pub fn unsigned_numeric_literal(
    input: &mut TokenStream,
) -> ModalResult<Spanned<UnsignedNumericLiteral>> {
    unsigned_integer
        .map(UnsignedNumericLiteral::Integer)
        .spanned()
        .parse_next(input)
}

pub fn unsigned_integer(input: &mut TokenStream) -> ModalResult<Spanned<UnsignedInteger>> {
    dispatch! {any;
        &TokenKind::UnsignedDecimalInteger(integer) => empty
            .value(UnsignedInteger {
                kind: UnsignedIntegerKind::Decimal,
                integer: integer.into(),
            }),
        &TokenKind::UnsignedOctalInteger(integer) => empty
            .value(UnsignedInteger {
                kind: UnsignedIntegerKind::Octal,
                integer: integer.into(),
            }),
        &TokenKind::UnsignedHexInteger(integer) => empty
            .value(UnsignedInteger {
                kind: UnsignedIntegerKind::Hex,
                integer: integer.into(),
            }),
        &TokenKind::UnsignedBinaryInteger(integer) => empty
            .value(UnsignedInteger {
                kind: UnsignedIntegerKind::Binary,
                integer: integer.into(),
            }),
        _ => fail,
    }
    .spanned()
    .parse_next(input)
}

pub fn unsigned_decimal_integer(input: &mut TokenStream) -> ModalResult<Spanned<UnsignedInteger>> {
    winnow::token::any
        .verify_map(|token: &Token| match token.kind {
            TokenKind::UnsignedDecimalInteger(integer) => Some(UnsignedInteger {
                kind: UnsignedIntegerKind::Decimal,
                integer: integer.into(),
            }),
            _ => None,
        })
        .spanned()
        .parse_next(input)
}

pub fn boolean_literal(input: &mut TokenStream) -> ModalResult<Spanned<BooleanLiteral>> {
    dispatch! {any;
        TokenKind::True => empty.value(BooleanLiteral::True),
        TokenKind::False => empty.value(BooleanLiteral::False),
        TokenKind::Unknown => empty.value(BooleanLiteral::Unknown),
        _ => fail
    }
    .spanned()
    .parse_next(input)
}

pub fn general_parameter_reference(input: &mut TokenStream) -> ModalResult<Spanned<Ident>> {
    dispatch! {any;
        TokenKind::GeneralParameterReference(name) => {
            cut_err(empty.verify_map(|_| name.unescape()))
        },
        _ => fail,
    }
    .spanned()
    .parse_next(input)
}

pub fn substituted_parameter_reference(input: &mut TokenStream) -> ModalResult<Spanned<Ident>> {
    dispatch! {any;
        TokenKind::SubstitutedParameterReference(name) => {
            cut_err(empty.verify_map(|_| name.unescape()))
        },
        _ => fail,
    }
    .spanned()
    .parse_next(input)
}

pub fn character_string_literal(input: &mut TokenStream) -> ModalResult<Spanned<StringLiteral>> {
    dispatch! {any;
        TokenKind::SingleQuoted(quoted) | TokenKind::DoubleQuoted(quoted) => {
            cut_err(empty.verify_map(|_| {
                Some(StringLiteral {
                    kind: StringLiteralKind::Char,
                    literal: quoted.unescape()?,
                })
            }))
        },
        _ => fail
    }
    .spanned()
    .parse_next(input)
}

pub fn node_synonym(input: &mut TokenStream) -> ModalResult<()> {
    one_of((TokenKind::Node, TokenKind::Vertex))
        .void()
        .parse_next(input)
}

pub fn edge_synonym(input: &mut TokenStream) -> ModalResult<()> {
    one_of((TokenKind::Edge, TokenKind::Relationship))
        .void()
        .parse_next(input)
}

pub fn edges_synonym(input: &mut TokenStream) -> ModalResult<()> {
    one_of((TokenKind::Edges, TokenKind::Relationships))
        .void()
        .parse_next(input)
}

pub fn implies(input: &mut TokenStream) -> ModalResult<()> {
    one_of((TokenKind::Implies, TokenKind::RightDoubleArrow))
        .void()
        .parse_next(input)
}

#[cfg(all(test, feature = "serde"))]
mod tests {
    use insta::assert_yaml_snapshot;

    use super::*;
    use crate::parser::utils::parse;

    #[test]
    fn test_identifier_1() {
        let parsed = parse!(identifier, "abcd");
        assert_yaml_snapshot!(parsed);
    }

    #[test]
    fn test_identifier_2() {
        let parsed = parse!(identifier, "`abcd`");
        assert_yaml_snapshot!(parsed);
    }

    #[test]
    fn test_identifier_3() {
        let parsed = parse!(identifier, "\"abcd\"");
        assert_yaml_snapshot!(parsed);
    }

    #[test]
    fn test_boolean_literal_1() {
        let parsed = parse!(boolean_literal, "true");
        assert_yaml_snapshot!(parsed);
    }

    #[test]
    fn test_boolean_literal_2() {
        let parsed = parse!(boolean_literal, "false");
        assert_yaml_snapshot!(parsed);
    }

    #[test]
    fn test_boolean_literal_3() {
        let parsed = parse!(boolean_literal, "unknown");
        assert_yaml_snapshot!(parsed);
    }

    #[test]
    fn test_unsigned_literal_1() {
        let parsed = parse!(unsigned_literal, "false");
        assert_yaml_snapshot!(parsed);
    }

    #[test]
    fn test_unsigned_literal_2() {
        let parsed = parse!(unsigned_literal, "null");
        assert_yaml_snapshot!(parsed);
    }

    #[test]
    fn test_character_string_literal() {
        let parsed = parse!(character_string_literal, "\"ab\\ncd\"");
        assert_yaml_snapshot!(parsed);
    }

    #[test]
    fn test_unsigned_integer_1() {
        let parsed = parse!(unsigned_integer, "123");
        assert_yaml_snapshot!(parsed);
    }

    #[test]
    fn test_unsigned_integer_2() {
        let parsed = parse!(unsigned_integer, "0o123");
        assert_yaml_snapshot!(parsed);
    }

    #[test]
    fn test_unsigned_integer_3() {
        let parsed = parse!(unsigned_integer, "0x123");
        assert_yaml_snapshot!(parsed);
    }

    #[test]
    fn test_unsigned_integer_4() {
        let parsed = parse!(unsigned_integer, "0b101");
        assert_yaml_snapshot!(parsed);
    }

    #[test]
    fn test_general_parameter_reference_1() {
        let parsed = parse!(general_parameter_reference, "$abc");
        assert_yaml_snapshot!(parsed);
    }

    #[test]
    fn test_general_parameter_reference_2() {
        let parsed = parse!(general_parameter_reference, "$\"abc\"");
        assert_yaml_snapshot!(parsed);
    }

    #[test]
    fn test_substituted_parameter_reference_1() {
        let parsed = parse!(substituted_parameter_reference, "$$abc");
        assert_yaml_snapshot!(parsed);
    }

    #[test]
    fn test_substituted_parameter_reference_2() {
        let parsed = parse!(substituted_parameter_reference, "$$\"abc\"");
        assert_yaml_snapshot!(parsed);
    }
}
