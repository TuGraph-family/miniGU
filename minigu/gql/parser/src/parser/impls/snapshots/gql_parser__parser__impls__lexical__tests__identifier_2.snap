---
source: minigu/parser/src/parser/impls/lexical.rs
expression: parsed
---
- abcd
- start: 0
  end: 6
